cmd/soroban-cli/src/utils.rs

```diff
@@ -196,77 +198,76 @@ pub fn contract_id_from_str(contract_id: &str) -> Result<[u8; 32], stellar_strke
         .map_err(|_| stellar_strkey::DecodeError::Invalid)
 }
 
+fn get_entry_from_spanshot(
+    key: &LedgerKey,
+    entries: &LedgerSnapshotEntries,
+) -> Option<(Box<LedgerEntry>, Option<u32>)> {
+    for (k, result) in entries {
+        if *key == **k {
+            return Some((*result).clone());
+        }
+    }
+    None
+}
+
 /// # Errors
 ///
 /// Might return an error
-pub fn get_contract_spec_from_storage(
-    storage: &mut Storage,
-    current_ledger_seq: &u32,
+pub fn get_contract_spec_from_state(
+    state: &LedgerSnapshot,
     contract_id: [u8; 32],
 ) -> Result<Vec<ScSpecEntry>, FromWasmError> {
+    let current_ledger_seq = state.sequence_number;
     let key = LedgerKey::ContractData(LedgerKeyContractData {
         contract: ScAddress::Contract(contract_id.into()),
         key: ScVal::LedgerKeyContractInstance,
-        body_type: ContractEntryBodyType::DataEntry,
         durability: ContractDataDurability::Persistent,
     });
-    match storage.get(&key.into(), &Budget::default()) {
-        Ok(rc) => match rc.as_ref() {
-            LedgerEntry {
-                data:
-                    LedgerEntryData::ContractData(ContractDataEntry {
-                        body:
-                            ContractDataEntryBody::DataEntry(ContractDataEntryData {
-                                val: ScVal::ContractInstance(ScContractInstance { executable, .. }),
-                                ..
-                            }),
-                        expiration_ledger_seq,
-                        ..
-                    }),
-                ..
-            } => match executable {
-                ContractExecutable::Token => {
-                    if expiration_ledger_seq <= current_ledger_seq {
-                        return Err(FromWasmError::NotFound);
-                    }
-                    let res = soroban_spec::read::parse_raw(&token::StellarAssetSpec::spec_xdr());
-                    res.map_err(FromWasmError::Parse)
+    let (entry, expiration_ledger_seq) = match get_entry_from_spanshot(&key, &state.ledger_entries)
+    {
+        // It's a contract data entry, so it should have an expiration if present
+        Some((entry, expiration)) => (entry, expiration.unwrap()),
+        None => return Err(FromWasmError::NotFound),
+    };
+    if expiration_ledger_seq <= current_ledger_seq {
+        return Err(FromWasmError::NotFound);
+    }
+
+    match *entry {
+        LedgerEntry {
+            data:
+                LedgerEntryData::ContractData(ContractDataEntry {
+                    val: ScVal::ContractInstance(ScContractInstance { executable, .. }),
+                    ..
+                }),
+            ..
+        } => match executable {
+            ContractExecutable::Token => {
+                // TODO/FIXME: I don't think it will work for token contracts, since we don't store them in the state?
+                let res = soroban_spec::read::parse_raw(&token::StellarAssetSpec::spec_xdr());
```

Comment:
The spec doesn't come from the state, just the entry that points to the token contract so this does work.

Reply:
Yes, I know the spec is obtained separately. What I meant is that, AFAIU, we don't store `ContractInstance` entries for token contracts since the user provides no wasm.

Maybe I am wrong, though, if that's the case, can you point me where that happens?

Reply:
A non Wasm contract instance still needs to store the fact that it's a token.

Reply:
> A non Wasm contract instance still needs to store the fact that it's a token.

Yes, but my point is that we only seem to update the state when a wasm file is provided, which doesn't for token contracts.